homeassistant/components/mqtt_room/sensor.py

```diff
@@ -118,48 +139,59 @@ def message_received(msg):
 
             device = _parse_update_data(msg.topic, data)
             if device.get(CONF_DEVICE_ID) == self._device_id:
-                if self._distance is None or self._updated is None:
-                    update_state(**device)
-                else:
+                for rec in self._room_states:
+                    if rec["distance"] is None or rec["updated"] is None:
+                        update_state(rec, **device)
+                        break
+
                     # update if:
                     # device is in the same room OR
                     # device is closer to another room OR
                     # last update from other room was too long ago
-                    timediff = dt.utcnow() - self._updated
+                    timediff = dt.utcnow() - rec["updated"]
                     if (
-                        device.get(ATTR_ROOM) == self._state
-                        or device.get(ATTR_DISTANCE) < self._distance
+                        device.get(ATTR_ROOM) == rec["room"]
+                        or device.get(ATTR_DISTANCE) < rec["distance"]
                         or timediff.total_seconds() >= self._timeout
                     ):
-                        update_state(**device)
+                        update_state(rec, **device)
+                        break
 
-        return await mqtt.async_subscribe(
-            self.hass, self._state_topic, message_received, 1
-        )
+        await mqtt.async_subscribe(self.hass, self._state_topic, message_received, 1)
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Return the name of the sensor."""
         return self._name
 
     @property
     def extra_state_attributes(self):
         """Return the state attributes."""
-        return {ATTR_DISTANCE: self._distance}
```

Comment:
I thought about just appending the additional room states to the dict directly using the room name instead of creating a sub-key "extra_rooms", but decided against it primarily because I cannot guarantee that the remaining namespace of the attributes won't clash with the user chosen names of the rooms/base stations. I.e. if I have a base station called "distance", it will override the distance key in the dict. We'd have to deal with potential collisions of any additional attributes that may get added in the future, too (e.g. raw signal strength, just as a reasonable example). Using a sub-key just seemed more resilient.

I have also considered multiple entities, but this approach frankly was much simpler and was within my capability of getting done in a reasonable amount of time. I'm a HA user first and have no real experience working in Python and on the Home Assistant code base. This solution seemed reasonable and met my own needs.

Reply:
Okay, I agree sticking to a separate attribute for now. To make this integration future proof it should get support for discovery. To get this working I could give some advice. Further this integration still sets up using `configuration.yaml` With deprecating `device_id` config option and implementing discovery we could make this integration more future proof.